File: hono-api/src/services/application.service.ts
import { eq, and, sql, asc, desc, gte, inArray } from 'drizzle-orm';
import { db } from '../db/client.js';
import { applications, interviewStages, type Application } from '../db/schema.js';
import type {
  CreateApplicationInput,
  UpdateApplicationInput,
  ListApplicationsQuery,
  ApplicationResponse,
  PaginatedApplicationsResponse,
} from '../types/api.js';
import { toApplicationResponse } from './shared.js';
import { recordHistory, buildDescription } from './history.service.js';

const FIELD_LABELS_MAP: Record<string, string> = {
  companyName: 'Company Name',
  positionTitle: 'Position Title',
  dateApplied: 'Date Applied',
  status: 'Status',
  companyUrl: 'Company URL',
  jobPostingUrl: 'Job Posting URL',
  companyCareerUrl: 'Career Page URL',
  companyCategory: 'Company Category',
  skillsMatch: 'Skills Match',
  jobSource: 'Job Source',
  coverLetterRequired: 'Cover Letter Required',
  specialRequirements: 'Special Requirements',
  salaryMin: 'Min Salary',
  salaryMax: 'Max Salary',
  notes: 'Notes',
  offerDueDate: 'Offer Due Date',
};

export async function listApplications(query: ListApplicationsQuery): Promise<PaginatedApplicationsResponse> {
  const { status, companyCategory, jobSource, skillsMatchMin, includeArchived, sortBy, sortDir, page, limit } = query;

  // Build conditions
  const conditions = [];

  if (!includeArchived) {
    conditions.push(eq(applications.isArchived, false));
  }

  if (status) {
    const statuses = status.split(',').map((s) => s.trim()) as Application['status'][];
    conditions.push(inArray(applications.status, statuses));
  }

  if (companyCategory) {
    conditions.push(eq(applications.companyCategory, companyCategory));
  }

  if (jobSource) {
    conditions.push(eq(applications.jobSource, jobSource));
  }

  if (skillsMatchMin) {
    conditions.push(gte(applications.skillsMatch, skillsMatchMin));
  }

  // Build order
  const sortColumn = {
    dateApplied: applications.dateApplied,
    companyName: applications.companyName,
    updatedAt: applications.updatedAt,
  }[sortBy];

  const orderFn = sortDir === 'asc' ? asc : desc;

  // Get total count
  const countResult = await db
    .select({ count: sql<number>`count(*)` })
    .from(applications)
    .where(conditions.length > 0 ? and(...conditions) : undefined);

  const total = Number(countResult[0]?.count || 0);

  // Get paginated results
  const offset = (page - 1) * limit;
  const apps = await db.query.applications.findMany({
    where: conditions.length > 0 ? and(...conditions) : undefined,
    orderBy: [orderFn(sortColumn)],
    limit,
    offset,
    with: {
      interviewStages: true,
    },
  });

  return {
    items: apps.map((app) => toApplicationResponse(app, app.interviewStages)),
    page,
    limit,
    total,
  };
}

export async function getApplication(id: string): Promise<ApplicationResponse | null> {
  const app = await db.query.applications.findFirst({
    where: eq(applications.id, id),
    with: {
      interviewStages: true,
    },
  });

  if (!app) return null;
  return toApplicationResponse(app, app.interviewStages);
}

export async function createApplication(input: CreateApplicationInput): Promise<ApplicationResponse> {
  const dateApplied = input.dateApplied || new Date().toISOString().split('T')[0];

  const [app] = await db
    .insert(applications)
    .values({
      companyName: input.companyName,
      positionTitle: input.positionTitle,
      dateApplied,
      companyUrl: input.companyUrl || null,
      jobPostingUrl: input.jobPostingUrl || null,
      companyCareerUrl: input.companyCareerUrl || null,
      companyCategory: input.companyCategory || null,
      skillsMatch: input.skillsMatch || null,
      jobSource: input.jobSource || null,
      coverLetterRequired: input.coverLetterRequired ?? null,
      specialRequirements: input.specialRequirements || null,
      salaryMin: input.salaryMin || null,
      salaryMax: input.salaryMax || null,
      notes: input.notes || null,
    })
    .returning();

  await recordHistory(app.id, buildDescription('create', `${app.companyName} - ${app.positionTitle}`));

  return toApplicationResponse(app, []);
}

export async function updateApplication(id: string, input: UpdateApplicationInput): Promise<ApplicationResponse | null> {
  // Check if exists
  const existing = await db.query.applications.findFirst({
    where: eq(applications.id, id),
  });

  if (!existing) return null;

  // Build update values
  const updateValues: Partial<typeof applications.$inferInsert> = {
    updatedAt: new Date(),
  };

  if (input.companyName !== undefined) updateValues.companyName = input.companyName;
  if (input.positionTitle !== undefined) updateValues.positionTitle = input.positionTitle;
  if (input.dateApplied !== undefined) updateValues.dateApplied = input.dateApplied;
  if (input.status !== undefined) updateValues.status = input.status;
  if (input.companyUrl !== undefined) updateValues.companyUrl = input.companyUrl;
  if (input.jobPostingUrl !== undefined) updateValues.jobPostingUrl = input.jobPostingUrl;
  if (input.companyCareerUrl !== undefined) updateValues.companyCareerUrl = input.companyCareerUrl;
  if (input.companyCategory !== undefined) updateValues.companyCategory = input.companyCategory;
  if (input.skillsMatch !== undefined) updateValues.skillsMatch = input.skillsMatch;
  if (input.jobSource !== undefined) updateValues.jobSource = input.jobSource;
  if (input.coverLetterRequired !== undefined) updateValues.coverLetterRequired = input.coverLetterRequired;
  if (input.specialRequirements !== undefined) updateValues.specialRequirements = input.specialRequirements;
  if (input.salaryMin !== undefined) updateValues.salaryMin = input.salaryMin;
  if (input.salaryMax !== undefined) updateValues.salaryMax = input.salaryMax;
  if (input.notes !== undefined) updateValues.notes = input.notes;
  if (input.offerDueDate !== undefined) updateValues.offerDueDate = input.offerDueDate;

  const [updated] = await db.update(applications).set(updateValues).where(eq(applications.id, id)).returning();

  const stages = await db.query.interviewStages.findMany({
    where: eq(interviewStages.applicationId, id),
  });

  // Record history after update (snapshot captures post-mutation state)
  const changedFields = Object.keys(input)
    .filter((key) => key in FIELD_LABELS_MAP)
    .map((key) => FIELD_LABELS_MAP[key]);
  if (changedFields.length > 0) {
    await recordHistory(id, buildDescription('update', changedFields.join(', ')));
  }

  return toApplicationResponse(updated, stages);
}

export async function deleteApplication(id: string): Promise<boolean> {
  await recordHistory(id, buildDescription('delete'));

  const result = await db.delete(applications).where(eq(applications.id, id)).returning({ id: applications.id });
  return result.length > 0;
}

export async function archiveApplication(id: string): Promise<ApplicationResponse | null> {
  const [updated] = await db
    .update(applications)
    .set({ isArchived: true, updatedAt: new Date() })
    .where(eq(applications.id, id))
    .returning();

  if (!updated) return null;

  const stages = await db.query.interviewStages.findMany({
    where: eq(interviewStages.applicationId, id),
  });

  await recordHistory(id, buildDescription('archive'));

  return toApplicationResponse(updated, stages);
}

export async function restoreApplication(id: string): Promise<ApplicationResponse | null> {
  const [updated] = await db
    .update(applications)
    .set({ isArchived: false, updatedAt: new Date() })
    .where(eq(applications.id, id))
    .returning();

  if (!updated) return null;

  const stages = await db.query.interviewStages.findMany({
    where: eq(interviewStages.applicationId, id),
  });

  await recordHistory(id, buildDescription('restore'));

  return toApplicationResponse(updated, stages);
}

File: rails-api/app/services/application_service.rb
class ApplicationService
  DEFAULT_STAGES = [
    "Phone Screen",
    "Technical Screen",
    "Coding Challenge",
    "Onsite/Panel",
    "Behavioral/Cultural Fit",
    "Final Decision"
  ].freeze

  FIELD_LABELS = {
    company_name: "Company Name",
    position_title: "Position Title",
    date_applied: "Date Applied",
    status: "Status",
    company_url: "Company URL",
    job_posting_url: "Job Posting URL",
    company_career_url: "Career Page URL",
    company_category: "Company Category",
    skills_match: "Skills Match",
    job_source: "Job Source",
    cover_letter_required: "Cover Letter Required",
    special_requirements: "Special Requirements",
    salary_min: "Min Salary",
    salary_max: "Max Salary",
    notes: "Notes",
    offer_due_date: "Offer Due Date",
    is_archived: "Archived"
  }.freeze

  SORT_COLUMNS = {
    "dateApplied" => :date_applied,
    "companyName" => :company_name,
    "status" => :status,
    "updatedAt" => :updated_at
  }.freeze

  def self.list(params)
    scope = JobApplication.includes(:interview_stages)
    scope = scope.where(is_archived: false) unless truthy?(params[:includeArchived])
    scope = scope.where(status: params[:status].to_s.split(",").map(&:strip)) if params[:status].present?
    scope = scope.where(company_category: params[:companyCategory]) if params[:companyCategory].present?
    scope = scope.where(job_source: params[:jobSource]) if params[:jobSource].present?
    if params[:skillsMatchMin].present?
      skills_match_min = Integer(params[:skillsMatchMin], exception: false)
      scope = scope.where(skills_match: skills_match_min..) if skills_match_min
    end

    total = scope.count
    page = [params.fetch(:page, 1).to_i, 1].max
    limit = params.fetch(:limit, 20).to_i.clamp(1, 100)
    sort_column = SORT_COLUMNS.fetch(params[:sortBy].to_s, :updated_at)
    sort_direction = params[:sortDir].to_s.downcase == "asc" ? :asc : :desc
    order_fragment = "#{JobApplication.connection.quote_column_name(sort_column)} #{sort_direction.to_s.upcase}"
    order_fragment = "#{order_fragment} NULLS LAST" if sort_column == :date_applied

    items = scope.order(Arel.sql(order_fragment)).offset((page - 1) * limit).limit(limit).map do |application|
      ApplicationSerializer.application(application)
    end

    { items:, page:, limit:, total: }
  end

  def self.create(payload)
    attributes = ApiParams.application_attributes(payload)
    attributes[:status] = attributes[:status].presence || "unsubmitted"
    apply_date_side_effects(attributes, on_create: true)

    application = nil
    ApplicationRecord.transaction do
      application = JobApplication.create!(attributes)
      create_default_stages_if_needed(application)
      ApplicationSnapshotService.record(application, "Created application")
    end
    application.reload
  end

  def self.update(application, payload)
    attributes = ApiParams.application_attributes(payload)
    return application if attributes.empty?

    previous_status = application.status
    prevent_terminal_transition!(application, attributes[:status])
    apply_date_side_effects(attributes, on_create: false)

    ApplicationRecord.transaction do
      application.update!(attributes)
      create_default_stages_if_needed(application) if previous_status != application.status
      ApplicationSnapshotService.record(application, update_description(attributes.keys))
    end
    application.reload
  end

  def self.archive(application)
    ApplicationRecord.transaction do
      application.update!(is_archived: true)
      ApplicationSnapshotService.record(application, "Archived application")
    end
    application.reload
  end

  def self.restore(application)
    ApplicationRecord.transaction do
      application.update!(is_archived: false)
      ApplicationSnapshotService.record(application, "Restored application")
    end
    application.reload
  end

  def self.delete(application)
    ApplicationRecord.transaction do
      ApplicationSnapshotService.record(application, "Deleted application")
      application.destroy!
    end
  end

  def self.truthy?(value)
    value == true || value.to_s.downcase == "true"
  end

  def self.apply_date_side_effects(attributes, on_create:)
    status = attributes[:status]
    return if status.blank?

    if status == "unsubmitted"
      attributes[:date_applied] = nil
    elsif on_create && (!attributes.key?(:date_applied) || attributes[:date_applied].nil?)
      attributes[:date_applied] = Date.current
    end
  end

  def self.prevent_terminal_transition!(application, new_status)
    return if new_status.blank? || application.status == new_status
    return unless JobApplication::TERMINAL_STATUSES.include?(application.status)

    application.errors.add(:status, "cannot transition from terminal status")
    raise ActiveRecord::RecordInvalid, application
  end

  def self.create_default_stages_if_needed(application)
    return unless application.status == "interviewing"
    return if application.interview_stages.exists?

    DEFAULT_STAGES.each_with_index do |name, index|
      application.interview_stages.create!(name: name, order: index + 1)
    end
  end

  def self.update_description(fields)
    labels = fields.filter_map { |field| FIELD_LABELS[field] }
    labels.empty? ? "Updated application" : "Updated: #{labels.join(', ')}"
  end
end
